### PR TITLE
Add `TxPointer` to input coin and contract

### PR DESCRIPTION
**Pull request by vlopes11**
related: https://github.com/FuelLabs/fuel-specs/pull/275

### CODE REVIEW

File: src/transaction/id.rs [hunk 1]
```diff
@@ -1,74 +1,76 @@
...
                 state_root,
+                tx_pointer,
                 ..
-            } = input
-            {
-                *utxo_id = UtxoId::default();
-                balance_root.iter_mut().for_each(|b| *b = 0);
-                state_root.iter_mut().for_each(|b| *b = 0);
+            } => {
+                mem::take(utxo_id);
```
Comment (Dentosal) [inline src/transaction/id.rs:25]: nit: when not using the return value of `mem::take`, I find using `*value = Default::default();` to be more readable

File: src/transaction/types/tx_pointer.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if buf.len() < Self::LEN {
+            return Err(bytes::eof());
+        }
+
+        // Safety: buf len is checked
+        let (block_height, buf) = unsafe { bytes::restore_word_unchecked(buf) };
+        let (tx_index, _) = unsafe { bytes::restore_word_unchecked(buf) };
+
+        self.block_height = block_height as u32;
+        self.tx_index = tx_index as u16;
```
Comment (Dentosal) [inline src/transaction/types/tx_pointer.rs:100]: Is there a possibility that these `as` casts are sinlently truncating?

File: tests/offset.rs [hunk 3]
```diff
@@ -35,9 +35,9 @@ fn tx_offset() {
     let mut tested_output_contract_created_id = false;
     let mut tested_output_recipient = false;
 
-    let cases = 3;
+    let cases = 2;
 
-    TransactionFactory::from_seed(1296)
+    TransactionFactory::from_seed(1295)
```
Comment (Dentosal) [inline tests/offset.rs:40]: Why are these changed?

Comment (vlopes11) [reply tests/offset.rs:40]: That is a magic touch indeed, but its just that the factory with a seed `1295` will result in only two transactions covering all the specified cases

File: src/transaction/types/tx_pointer.rs (hunk 2 above)
Comment (vlopes11) [reply src/transaction/types/tx_pointer.rs]: The original thought is that it might be overkill to check that since block height is `u32` everywhere in the stack, as `tx_index` is `u16`. Our canonical serialization only encode/decode words/`u64`, so we need to truncate

But you're right; better safe than sorry: https://github.com/FuelLabs/fuel-tx/pull/170/commits/1368be316e15c67e01a9fb43fff7f4324b103fc9

File: src/transaction/id.rs (hunk 1 above)
Comment (vlopes11) [reply src/transaction/id.rs:25]: isn't the shorter the better? I like `mem::take(utxo_id);` because of the reduced verbosity, and the std api is clear on what it is doing (e.g. taking the value)

Comment (Dentosal) [reply src/transaction/id.rs:25]: Explict is better, `take` is intended for a specific situation. Docs are clear, yes, but with =default I wouldn't have to check the docs at all. Anyway, if you feel like this is better, we can just go with it.

File: tests/offset.rs (hunk 3 above)
Comment (Dentosal) [reply tests/offset.rs:40]: Maybe add a comment explaining that?

Comment (vlopes11) [reply tests/offset.rs]: https://github.com/FuelLabs/fuel-tx/pull/170/commits/d62afa10d3ba2de8f6496eca42ba0674c21ccedc

**Review by Dentosal — APPROVED** [review]
